internal_types/src/entry.rs

```diff
@@ -1079,6 +1087,341 @@ enum ColumnRaw<'a> {
+                    } ).collect::<Result<Vec<_>>>()?;
+                let partition_writes = fbb.create_vector(&partition_writes);
+
+                let write_operations = entry_fb::WriteOperations::create(
+                    &mut fbb,
+                    &entry_fb::WriteOperationsArgs {
+                        partition_writes: Some(partition_writes),
+                    },
+                );
+
+                entry_fb::Entry::create(
+                    &mut fbb,
+                    &entry_fb::EntryArgs {
+                        operation_type: entry_fb::Operation::write,
+                        operation: Some(write_operations.as_union_value()),
+                    },
+                )
+            }
+            _ => panic!("unsupported"),
+        };
+
+        let sequenced_entry = entry_fb::SequencedEntry::create(
+            &mut fbb,
+            &entry_fb::SequencedEntryArgs {
+                clock_value,
+                writer_id,
+                entry: Some(entry),
+                entry_bytes: None,
+            },
+        );
+
+        fbb.finish(sequenced_entry, None);
+
+        let (mut data, idx) = fbb.collapse();
+        let sequenced_entry = Self::try_from(data.split_off(idx))
+            .expect("Flatbuffer data just constructed should be valid");
+
+        Ok(sequenced_entry)
+    }
+
+    /// Returns the Flatbuffers struct for the SequencedEntry
+    pub fn fb(&self) -> &entry_fb::SequencedEntry<'_> {
+        self.borrow_fb()
+    }
+
+    pub fn partition_writes(&self) -> Option<Vec<PartitionWrite<'_>>> {
+        match self.fb().entry().as_ref() {
+            Some(e) => match e.operation_as_write().as_ref() {
+                Some(w) => w
+                    .partition_writes()
+                    .as_ref()
+                    .map(|w| w.iter().map(|fb| PartitionWrite { fb }).collect::<Vec<_>>()),
+                None => None,
+            },
+            None => None,
+        }
+    }
+
+    pub fn clock_value(&self) -> u64 {
+        self.fb().clock_value()
+    }
+
+    pub fn writer_id(&self) -> u32 {
+        self.fb().writer_id()
+    }
+}
+
+impl TryFrom<Vec<u8>> for SequencedEntry {
+    type Error = flatbuffers::InvalidFlatbuffer;
+
+    fn try_from(data: Vec<u8>) -> Result<Self, Self::Error> {
+        SequencedEntryTryBuilder {
+            data,
+            fb_builder: |data| flatbuffers::root::<entry_fb::SequencedEntry<'_>>(data),
+        }
+        .try_build()
+    }
+}
+
+#[self_referencing]
+#[derive(Debug)]
+pub struct SequencedEntryRaw {
+    data: Vec<u8>,
+    #[borrows(data)]
+    #[covariant]
+    fb: entry_fb::SequencedEntry<'this>,
+    #[borrows(data)]
+    #[covariant]
+    entry: Option<entry_fb::Entry<'this>>,
+}
+
+impl SequencedEntryRaw {
+    pub fn new_from_entry_bytes(
+        clock_value: u64,
+        writer_id: u32,
+        entry_bytes: &[u8],
+    ) -> Result<Self> {
+        let mut fbb = FlatBufferBuilder::new_with_capacity(1024);
```

Comment:
Can this be allocated relative to the input to avoid reallocations?

```suggestion
        // The flatbuffer contains:
        //    1xu64 -> clock_value
        //    1xu32 -> writer_id
        //    0?       -> entry (unused here)
        //    input   -> entry_bytes
        // The buffer also needs space for the flatbuffer vtable.
        const OVERHEAD: usize = 4 * std::mem::size_of::<u64>();
        let mut fbb = FlatBufferBuilder::new_with_capacity(entry_bytes.len() + OVERHEAD);
```

I guesstimated the overhead from reading the rust types, so may be a bit off! A test to assert the size of the output is not greater than `input + OVERHEAD` would future-proof any changes causing reallocations because of a bad `OVERHEAD` value.

Reply:
I just tested this - seems to be fine 👍

Reply:
vtable for a table of N fields is: `(N+2)*sizeof(u16)`, in our case 12 bytes.

entry_bytes is encoded as a u32 offset + the body of the array which is a u32 size prefix plus the actual data.

the entry field is unused but still takes the space of a u32 offset.

overhead = `(4+2) * sizeof(u16) + sizeof(u64) + sizeof(u32) + sizeof(u32) + 2*sizeof(u32)` = 36 bytes.

once `entry` is removed, the overhead would be exactly 32 bytes (matching the `4*sizeof(u64)` here).

Perhaps we should mention in the comment that this is not just the vtable but also other other encoding overheads?

Reply:
I forgot: the flatbuffer starts with a 32 bit offset to the root object: so that's 40 bytes (or 36 once we remove the old `entry`). The underlying allocator will probably round up the allocation to the closest bucket size anyway, so we may not notice any difference depending on the size of entry_bytes)

Reply:
Experimentally I found 2 * u64 was fine but kept it as 4 to play it safe.

I don't think the exact amount of overhead is actually important, but allocating enough is! 😆